File: dpendulum-lagrange.jl
using DifferentialEquations
using Plots

function dpendulum(du, u, p, t)
    sdiff = sin(u[1] - u[2])
    cdiff = cos(u[1] - u[2])
    detM = 2 - cdiff^2
    du[1] = u[3]
    du[2] = u[4]
    du[3] = (sin(u[2])*cdiff - 2*sin(u[1]) - sdiff*(u[4]^2 + cdiff*u[3]^2)) / detM
    du[4] = (2*sin(u[1])*cdiff -2*sin(u[2]) + sdiff*(2*u[3]^2 + cdiff*u[4]^2)) / detM
end

function energy(u)
    e = u[3]^2 + (1/2)*u[4]^2 + u[3]*u[4]*cos(u[1] - u[2]) + 3 - 2*cos(u[1]) - cos(u[2])
end

u0 = [0.0, pi/2, 0.0, 0.0]

t_max = 384.0
t_step = 0.02 
prob = ODEProblem(dpendulum, u0, (0.0, t_max))
sol = solve(prob, Vern7(), adaptive=false, dt=t_step)

#display(sol)
data = reduce(hcat, sol.u)'

gr(size = (300, 300))
Plots.scalefontsizes()
Plots.scalefontsizes(0.4)
plotOrbit = scatter(
    data[:,1], data[:,3],
    aspect_ratio = 1,
    xlabel = "\\theta_1",
    ylabel = "\\nu_1",
    plot_title = "Double pendulum",
    markersize = 0.5,
    alpha = 0.25,
    legend = false,
    dpi = 300
)

annotate!(
    xlims(plotOrbit)[1] + 0.12*(xlims(plotOrbit)[2] - xlims(plotOrbit)[1]),
    ylims(plotOrbit)[2] - 0.03*(ylims(plotOrbit)[2] - ylims(plotOrbit)[1]),
    text("E = $(round(energy(u0), digits=4))", :red, :right, 4)
)
savefig(plotOrbit, "orbit-lagrange.png")

File: ode/double_pendulum/python/dpendulum-lagrange.py
import numpy as np
import matplotlib.pyplot as plt

from numpy import sin, cos
from scipy.integrate import solve_ivp

t_max = 2*192.0
dt = 0.02

x_init = [0, np.pi/2, 0, 0]

def energy(x):
    theta_1 = x[0]
    theta_2 = x[1]
    u_1 = x[2]   # \dot{\theta_1}
    u_2 = x[3]   # \dot{\theta_2}
    E = u_1**2 + (1/2)*u_2**2 + u_1*u_2*cos(theta_1 - theta_2) + 3 - 2*cos(theta_1) - cos(theta_2)
    return E

def dfunc(t, x):
    theta_1 = x[0]
    theta_2 = x[1]
    u_1 = x[2]   # \dot{\theta_1}
    u_2 = x[3]   # \dot{\theta_2}
    sdiff = sin(theta_1 - theta_2)
    cdiff = cos(theta_1 - theta_2)
    detM = 2 - cdiff**2
 
    res = np.zeros_like(x)
    res[0] = u_1
    res[1] = u_2
    res[2] = (sin(theta_2)*cdiff - 2*sin(theta_1) - sdiff*(u_2**2 + cdiff*u_1**2)) / detM
    res[3] = (2*sin(theta_1)*cdiff -2*sin(theta_2) + sdiff*(2*u_1**2 + cdiff*u_2**2)) / detM
    return res

e0 = energy(x_init)
sol = solve_ivp(dfunc, [0, t_max], x_init, t_eval=np.arange(0, t_max, dt), method="DOP853", rtol=1e-10, atol=1e-10)

fig = plt.figure(figsize=(8, 8))
plt.title("Double pendulum")
plt.axis("equal")
plt.xlabel("theta_1")
plt.ylabel("u_1")
plt.grid(True)

plt.plot(sol.y[0], sol.y[2], fillstyle='none', marker='o', markersize=1.0, linestyle='None', alpha=0.25)
xmin, xmax, ymin, ymax = plt.axis()
plt.text(xmin + 0.05*(xmax - xmin), ymax - 0.03*(xmax - xmin), "E = {}".format(e0))
fig.savefig('orbit-lagrange.png', dpi=150)

fig = plt.figure(figsize=(8, 8))
plt.title("Double pendulum. Invariant error")
plt.xlabel("Time")
plt.ylabel("Error")
plt.grid(True)

plt.plot(sol.t, energy(sol.y) - e0)
plt.legend(['Energy'])
fig.savefig('error-lagrange.png', dpi=150)
